graphql_compiler/interpreter/debugging.py

```diff
@@ -0,0 +1,186 @@
+    - actuation of all input and output iterables.
+
+    Traces may record complete executions of queries (i.e. ones where the generators are run
+    to completion), and may also be used to record partial executions (e.g. "fetch only 3 results").
+
+    The recording is sufficiently detailed to be used in at least the following use cases:
+    - to examine, either manually or through automated tests, a query's execution flow;
+    - to "impersonate" an InterpreterAdapter implementation for a given query, allowing bugs in
+      the interpreter library code to be isolated and reproduced without an InterpreterAdapter;
+    - to "impersonate" the interpreter library code toward an InterpreterAdapter implementation,
+      allowing adapter implementers to test and examine their implementations in isolation from
+      the rest of the interpreter library code.
+    """
+
+    DEFAULT_ROOT_UID: ClassVar[int] = -1
+
+    # AdapterOperations that are part of this trace but do not have a specific parent operation
+    # instead get assigned the trace's root_uid value as their parent_uid value.
+    root_uid: int = field(init=False, default=DEFAULT_ROOT_UID)
+    operations: Tuple[AdapterOperation, ...]
+
+
+class TraceRecorder(Generic[DataToken]):
+
+    # We expose an immutable (copied) version of the operation log through get_trace().
+    # Other attributes are considered public.
+    _operation_log: List[AdapterOperation]
+    root_uid: int
+
+    def __init__(self) -> None:
+        """Initialize the TraceRecorder."""
+        self._operation_log = []
+        self.root_uid = RecordedTrace[DataToken].DEFAULT_ROOT_UID
+
+    def record_call(
+        self,
+        operation_name: str,
+        parent_uid: int,
+        call_args: Tuple[Any, ...],
+        call_kwargs: Dict[str, Any],
+    ) -> int:
+        """Record that a call of the specified function has occurred with the given arguments."""
+        uid = len(self._operation_log)
+        call_args = deepcopy(call_args)
+        call_kwargs = deepcopy(call_kwargs)
+        self._operation_log.append(
+            AdapterOperation("call", operation_name, uid, parent_uid, (call_args, call_kwargs))
+        )
+        return uid
+
+    def record_iterable(
+        self, operation_name: str, parent_uid: int, iterable: Iterable[T]
+    ) -> Iterable[Tuple[int, T]]:
+        """Record each data item in the given iterable as a yield with the provided metadata.
+
+        Args:
+            operation_name: name to associate with this operation; usually the name of the function
+                            whose iterable output is being recorded.
+            parent_uid: unique identifier of the parent operation of this operation.
+            iterable: iterable to record
+
+        Yields:
+            for each piece of data yielded by the input iterable, yields a tuple of the
+            operation's unique identifier for that data together with the data itself
```

Comment:
sentence is a bit wordy. solution might be to do `yields a tuple of:` and then a bulleted list with the elements of the tuple. comes up in `record_compound_iterable` as well.

Reply:
Tweaked as you suggested, let me know if this is better.